pkg/storage/gcs.go

```diff
@@ -0,0 +1,140 @@
+)
+
+const (
+	gcsEndpointOption     = "gcs.endpoint"
+	gcsStorageClassOption = "gcs.storage_class"
+	gcsPredefinedAcl      = "gcs.predefined_acl"
+	gcsCredentialsFile    = "gcs.credentials_file"
+)
+
+// GCSBackendOptions are options for configuration the GCS storage.
+type GCSBackendOptions struct {
+	Endpoint        string `json:"endpoint" toml:"endpoint"`
+	StorageClass    string `json:"storage_class" toml:"storage_class"`
+	PredefinedAcl   string `json:"predefined_acl" toml:"predefined_acl"`
+	CredentialsFile string `json:"credentials_file" toml:"credentials_file"`
+}
+
+func (options *GCSBackendOptions) apply(gcs *backup.GCS) error {
+	if options.CredentialsFile == "" {
+		return errors.New("must provide 'gcs.credentials_file'")
+	}
+
+	gcs.Endpoint = options.Endpoint
+	gcs.StorageClass = options.StorageClass
+	gcs.PredefinedAcl = options.PredefinedAcl
+
+	b, err := ioutil.ReadFile(options.CredentialsFile)
+	if err != nil {
+		return err
+	}
+	gcs.CredentialsBlob = string(b)
+	return nil
+}
+
+func defineGCSFlags(flags *pflag.FlagSet) {
+	flags.String(gcsEndpointOption, "", "Set the GCS endpoint URL")
+	flags.String(gcsStorageClassOption, "", "Specify the GCS storage class for objects. If it is not set, objects uploaded are followed by the default storage class of the bucket. See https://cloud.google.com/storage/docs/storage-classes for valid values.")
+	flags.String(gcsPredefinedAcl, "", "Specify the GCS predefined acl for objects. If it is not set, objects uploaded are followed by the acl of bucket scope. See https://cloud.google.com/storage/docs/access-control/lists#predefined-acl for valid values.")
+	flags.String(gcsCredentialsFile, "", "Set the GCS credentials file path. You can get one from https://console.cloud.google.com/apis/credentials.")
+
+	_ = flags.MarkHidden(gcsEndpointOption)
+}
+
+func getBackendOptionsFromGCSFlags(flags *pflag.FlagSet) (options GCSBackendOptions, err error) {
+	options.Endpoint, err = flags.GetString(s3EndpointOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.StorageClass, err = flags.GetString(gcsStorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.PredefinedAcl, err = flags.GetString(gcsPredefinedAcl)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.CredentialsFile, err = flags.GetString(gcsCredentialsFile)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	return
+}
+
+type gcsStorage struct {
+	gcs    *backup.GCS
+	bucket *storage.BucketHandle
+}
+
+// Write file to storage
+func (s *gcsStorage) Write(ctx context.Context, name string, data []byte) error {
+	object := s.gcs.Prefix + name
+	wc := s.bucket.Object(object).NewWriter(ctx)
+	wc.StorageClass = s.gcs.StorageClass
+	wc.PredefinedACL = s.gcs.PredefinedAcl
+	_, err := wc.Write(data)
+	if err != nil {
+		return err
+	}
+	return wc.Close()
+}
+
+// Read storage file
+func (s *gcsStorage) Read(ctx context.Context, name string) ([]byte, error) {
```

Comment:
This reads the entire file into memory. However, at least some use cases (for example the checksum command) can stream data, avoiding the need to allocate a large amount of memory and finishing more quickly by starting the checksum calculation as soon as data is available. With reduced memory usage it is safer to parallelize the operation.

For this I think you can just add a function `Reader(Context, string) (io.Reader, error)` to the storage interface. Then the checksum command can use `io.Copy` to stream data through the checksum in blocks. It may be preferable to now have a `Read` function and instead return the Reader and size: this way the caller can read into memory if they prefer.

It is possible that the streaming approach could end up being slower, so it is definitely still worth benchmarking.

Reply:
The interface is definded using `[]byte` alreadly, see https://github.com/pingcap/br/blob/master/pkg/storage/storage.go#L9-L16
We may change the interface in another PR.

Another consideration is that `br` only backups metadata, which need to be hold in memory completly. So I think both interface in this situation is reasonable.

Reply:
I've created an issue #107, let's disscuss here.